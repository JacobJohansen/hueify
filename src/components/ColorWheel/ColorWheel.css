.color-overlay {
  position: absolute;
  top: 0px;
  left: 0px;
  width: 100%;
  height: 100%;
  display: flex;
  background: var(--body-bg);
  align-items: center;
  animation: fadeIn .1s ease-in;
}

.color-overlay .wheel {
  border-radius: 50%;
  margin-left: auto;
  margin-right: auto;
  animation: scaleIn .1s ease-in;
  cursor: pointer;
}

@keyframes fadeIn {
  from { opacity: 0; }
  to { opacity: 1; }
}

@keyframes scaleIn {
  from { opacity: 0; transform: rotate(45deg) scale(.7); }
  to { opacity: 1; transform: rotate(0deg) scale(1); }
}
