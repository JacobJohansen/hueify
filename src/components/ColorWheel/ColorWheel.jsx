/* global Image */
import { Component } from 'preact'
import './ColorWheel.css'

export default class ColorWheel extends Component {
  canvas = null
  size = 250
  center = null

  componenDidMount () {
    this.setupCanvas()
  }

  componentWillReceiveProps (nextProps) {
    this.props = nextProps
    this.setupCanvas()
  }

  componentWillUnmount () {
    this.canvas = null
  }

  setupCanvas () {
    if (!this.props.show) {
      return
    }

    process.nextTick(() => {
      this.canvas = document.querySelector('#canvas')
      this.center = [this.canvas.width / 2, this.canvas.height / 2]
      this.drawCircle()
      this.drawInnerCircle()
      this.convertImageToCanvas(this.convertCanvasToImage())
    })
  }

  drawCircle () {
    const context = this.canvas.getContext('2d')
    const [x, y] = this.center
    const radius = this.size / 2

    for (let angle = 0; angle <= 360; angle++) {
      const startAngle = (angle - 2) * Math.PI / 180
      const endAngle = angle * Math.PI / 180
      context.beginPath()
      context.moveTo(x, y)
      context.arc(x, y, radius, startAngle, endAngle)
      context.closePath()
      context.fillStyle = 'hsl(' + angle + ', 100%, 50%)'
      context.fill()
    }
  }

  drawInnerCircle () {
    const context = this.canvas.getContext('2d')
    const [x, y] = this.center
    const radius = this.size / 2

    const innerRadius = 10
    const outerRadius = this.size / 2

    const gradient = context.createRadialGradient(x, y, innerRadius, x, y, outerRadius)
    gradient.addColorStop(0, 'white')
    gradient.addColorStop(1, 'rgba(255, 255, 255, 0.2)')

    context.arc(x, y, radius, 0, 2 * Math.PI)

    context.fillStyle = gradient
    context.fill()
  }

  convertImageToCanvas (image) {
    this.canvas.getContext('2d').drawImage(image, 0, 0)
  }

  convertCanvasToImage () {
    const image = new Image()
    image.src = this.canvas.toDataURL('image/png')
    return image
  }

  onClick (event) {
    const { data } = this.canvas.getContext('2d').getImageData(
      event.offsetX,
      event.offsetY,
      1,
      1
    )

    if (typeof this.props.onChange === 'function') {
      this.props.onChange(data)
    }
  }

  close (event) {
    if (event.target.classList.contains('wheel')) {
      return
    }

    if (typeof this.props.onClose === 'function') {
      this.props.onClose()
    }
  }

  render ({ show }) {
    if (!show) {
      return null
    }

    return (
      <div className='color-overlay' onClick={this.close.bind(this)}>
        <canvas
          onClick={this.onClick.bind(this)}
          className='wheel'
          width={this.size}
          height={this.size}
          id='canvas'
        />
      </div>
    )
  }
}
