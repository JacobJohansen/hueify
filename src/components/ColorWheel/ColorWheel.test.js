/* eslint-env jest */
import { default as renderToString } from 'preact-render-to-string'
import { h, render } from 'preact' // eslint-disable-line
import ColorWheel from './ColorWheel'

global.h = h

beforeEach(() => {
  document.body.innerHTML = ''
})

test(`Shouldn't render anything if no props is set`, () => {
  const vdom = renderToString(<ColorWheel />)
  expect(vdom).toMatchSnapshot()
})

test(`Should render the color wheel if the show prop is set`, () => {
  const vdom = renderToString(<ColorWheel show />)
  expect(vdom).toMatchSnapshot()
})

test(`Should invoke the onClose method when the color wheel closes`, done => {
  const onClose = () => done()

  render(
    <ColorWheel show onClose={onClose} />,
    document.body
  )

  document.querySelector('.color-overlay').click()
})

test(`Should return color coordinates if color wheel is clicked`, done => {
  const onClick = rgb => {
    expect(rgb).toMatchSnapshot()
    done()
  }

  ColorWheel.prototype.onClick = function () {
    this.props.onClick([1, 2, 3])
  }

  render(
    <ColorWheel show onClick={onClick} />,
    document.body
  )

  document.querySelector('.wheel').click()
})
