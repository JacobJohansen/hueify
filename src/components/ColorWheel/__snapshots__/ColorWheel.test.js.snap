// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the color wheel if the show prop is set 1`] = `"<div class=\\"color-overlay\\"><canvas class=\\"wheel\\" width=\\"250\\" height=\\"250\\" id=\\"canvas\\"></canvas></div>"`;

exports[`Should return color coordinates if color wheel is clicked 1`] = `
Array [
  1,
  2,
  3,
]
`;

exports[`Shouldn't render anything if no props is set 1`] = `""`;
